kivy/lang/builder.py

```diff
@@ -581,8 +594,13 @@ def _apply_rule(self, widget, rule, rootrule, template_ctx=None,
                 # apply(), and so, we could use "self.parent".
                 child = cls(__no_builder=True)
                 widget.add_widget(child)
-                self.apply(child)
-                self._apply_rule(child, crule, rootrule)
+                self.apply(child, rule_children=rule_children)
+                child.dispatch('on_kv_applied')
+                self._apply_rule(
+                    child, crule, rootrule, rule_children=rule_children)
```

Comment:
Do you think it should perhaps be dispatched after `_apply_rule ` rather than before?

Reply:
I might misunderstand what the 'own rule' is. In the code below:

```python3
Builder.load_string('''
<WidgetA@Widget>:  # A
    ...            # A
<WidgetB@Widget>:
    WidgetA:       # B
        ...        # B
''')
root = Factory.WidgetB()
```

`WidgetA`'s own rule is `A` only, am I correct? If so, the current code should be fine. Because the following part applies `B`.

```python3
                self._apply_rule(
                    child, crule, rootrule, rule_children=rule_children)
```

Regardless of that, I probably should leave this PR to you. Because the PR really affects to the usability of Kivy, and misunderstanding your words is critical. So I want to leave this PR to you, if you don't mind.

Reply:
I only said `own rule` because that's what @tito said. But I doubt that he meant it like you understood it - English is not his first language, because to me it'll be confusing if it's fired before all the elements being applied to the widget is triggered and that includes `B` for me. 


I'm sorry that this took so long and that you had to re-write it so much. I think it was really helpful to arrive at a much better PR than either for us initially proposed. I didn't want to initially just take over the PR and do it myself as a competing PR, because I was hoping to illustrate, through comments, the thought and process that goes into API design so that future contributors can learn from it. So I'm really thankful for you patience.


Can I ask you to do one more thing? Can you please create a new branch locally, squash all the commits into it and open a new PR. That way you'll get the credit for you work on this because it'll be your commit, and then I can just polish it up from there by working on the PR. You can do it with something like this:

```
git checkout master
git checkout -b kv_events
git merge --squash post_kv
git commit -m "..."
git push origin kv_events
```